File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public boolean fun(int target, int[] diffs, int[] times, long limit) {
        long total = 0;
        for(int i = 0; i < diffs.length; i++) {
            if(diffs[i] <= target) {
                total += times[i];
            } else {
                int count = diffs[i] - target;
                total += (times[i - 1] + times[i]) * count + times[i];
            }
        }
        return total <= limit;
    }
    public int solution(int[] diffs, int[] times, long limit) {
        
        int start  = 1;
        int end = 100000;
        
        while(start < end) {
            int mid = (start + end) / 2;
            if(fun(mid, diffs, times, limit)) end = mid;
            else start = mid + 1;
        }
        return end;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    
    def is_valid(target: int) -> bool:
        count = 0
        for i in range(len(diffs)):
            if diffs[i] <= target:
                count += times[i]
            else:
                count += (times[i - 1] + times[i]) * (diffs[i] - target) + times[i]
            if count > limit:
                return False
        return True
    
    start, end = 1, 100000
    while start < end:
        mid = (start + end) // 2
        if is_valid(mid):
            end = mid
        else:
            start = mid + 1
    
    return start
